README.md
<a href="https://telatin.github.io/seqfu2"><img alt="seqfu logo" align="right" width="130" height="130" src="https://raw.githubusercontent.com/quadram-institute-bioscience/seqfu/master/docs/seqfu-logo-small.png"></p>

[![Unmaintained](https://img.shields.io/maintenance/no/2019?style=flat-square)](https://telatin.github.io/seqfu2)
[![Moved to SeqFu2](https://img.shields.io/badge/repository%20moved-seqfu-blue?style=flat-square)](https://telatin.github.io/seqfu2)
[![Perl module](https://img.shields.io/badge/Perl%20Module-Proch::N50-brown?style=flat-square)](https://metacpan.org/dist/Proch-N50)

➡️ A rewritten version of the suite is available: **[seqfu2](https://telatin.github.io/seqfu2)**


### Install SeqFu

SeqFu is available from Bioconda:

```
conda install -c bioconda seqfu
```

### `n50` utility and Perl module

The `Proch::N50` Perl module, and its utility `n50` are still maintained 
in this repository. Please open an [issue](https://github.com/quadram-institute-bioscience/seqfu/issues) here
if you find problems using either.

*n50* is available both via CPAN and Miniconda:
```
# Install via CPAN minus
cpanm Proch::N50

# Install via Miniconda
conda install -c bioconda n50
```

📦 See [Metacpan page](https://metacpan.org/dist/Proch-N50)

### Citation

Telatin A, Fariselli P, Birolo G. *SeqFu: A Suite of Utilities for the Robust and Reproducible Manipulation of Sequence Files*. 
Bioengineering 2021, 8, 59. [doi.org/10.3390/bioengineering8050059](https://doi.org/10.3390/bioengineering8050059)


n50/Changes
1.4.0   2021-08-05
        * Testing support for Windows newlines
        * Updating documentation 

1.3.0	2020-04-11
	* checking that the input files contains sequences
	* BUG FIX: Wrong auN returned if thousands separator was also enabled
	* 'n50' program now supports custom "Nx" (-e INT)

1.2.0	2020-04-09
  * getStats() now also returns "N75" and "N90" statistics
	* getStats() now also returns "auN" metric https://lh3.github.io/2020/04/08/a-new-metric-on-assembly-contiguity
	* getStats() now also returns "path", the absolute path to "dirname"
	* getStats() can return a N{X} metrics, if invoked as getStats(file, json, X);
	* 'n50' program partially supports the new metrics

1.0.0	2020-03-16
	* First stable release (stable APIs)
	* Switched to semantic versioning
	* Moved repository in the 'seqfu' package
	* FIXED sorting bug in 'n50'

0.92	2020-01-31
	* Downgrading requisite to perl 5.14

0.90	2020-01-16
	* FIXED: binary 'n50': Sorting tables did not work with -q (thousands separator)
	* FIXED: binary 'n50': Quotes added when using -q (thousands separator) and CSV format

0.83	2019-08-20
	* Moved 'n50' tests to xt/release

0.82	2019-08-19
	* Extending test suite on the 'n50' implementation
	* Thousand separator: Fixed a bug in 'n50' where the feature was always on

0.80	2019-08-18
	* Only "bin/n50" has been updated
	* NEW: Added -q for thousand separator (will cause adding quotes in --format csv)
	* Now also n50 migrated to JSON::PP, as Proch::N50 does since 0.04
	* Prints a short usage if invoked without filenames
	* Warning printed if using --format custom without --template
	* Added aliases in --template for newline ({n} and \n) and tab ({t} and \t)

0.80	2019-08-16
	* NEW: binary 'n50': Added -q for thousand separator
	* Now also n50 migrated to JSON::PP, as Proch::N50 does since 0.04
	* Prints a short usage if invoked without filenames
	* Warning printed if using --format custom without --template
	* Added aliases in --template for newline ({n} and \n) and tab ({t} and \t)

0.70	2019-08-16
	* FIXED: binary 'n50' was producing a pretty printed JSON even without --pretty [since transition to JSON:PP]

0.60	2019-08-15
	* MAJOR: Renamed the bundled script from n50.pl to n50
	* Minor cleanup
	* First release available in 'bioconda'

0.20	2019-06-30
	* Simplified test in JSON method (output unchanged)
	* Changes file cleaned; added a test for wide chars in changelog

0.10	2019-05-30
	* NEW: added sorting function to n50.pl (useful for 'screen' format)
	* FIXED: JSON object returned N50, min and max quoted (strings) rather than integers
	* Minor changes in dist.ini and release tests

0.09	2019-05-29
	* Added ./bin directory to provide excutable scripts
	* n50.pl now accepts sequence with zero length (as seqkit)
	* n50.pl new output format (screen friendly) using Text::ASCIITables only if requested
	* Coverage directives for Devel::Coverage
	* Extensive test of the algorithm

0.08	2019-05-23
	* Removed redundant POD paragraphs
	* Added unit test on non existing file
	* Fixed bug in n50.pl script
	* Added a test of the n50.pl script after build
	* Added a test of Changes that should contain last version

0.07	2019-05-22
	* PAUSE indexing problem
	* Updated FASTX::Reader to 0.09

0.06	2019-05-15
	* Moved version check to release tests

0.05	2019-05-09
	* MAJOR: Transition to FASTX::Reader to parse FASTA file
	* Code updated
	* Improving documentation coverage
	* Feature freeze aiming to release v 0.1 stable

0.04	2019-05-06
	* Transition to JSON::PP, now required since its a core module
	* Updated POD documentation
	* Updated JSON tests
	* Resubmitted as v 0.040 (PAUSE upload problem)

0.039	2019-05-05
	* NEW: Added 'min' and 'max' (lengths) in returned stats
	* Updated requirements specs in YML distribution file

0.032	2019-05-04
	* JSON unit test: restored conditional test
	* Removed unnecessary files from distribution (Dist::Zilla::Plugin::VersionFromModule/)

0.031	2019-05-03
	* Removed JSON unit test: this module has been built to only optionally require it. Even unit tests can have bugs.

0.03	2019-05-02
	* Added method _getJsonStats()_ to directly receive JSON string
	* Expanded json unit test accordingly

0.023	2019-05-01
	* Updated dist.ini [Dist::Zilla::Role::ModuleMetadata]
	* Synopsis uses print instead of say
	* Added CPAN distribution tests (CPAN Version check, POD integrity, Changes, Kwalitee...)
	* Added unit tests for getStats() and getStats with JSON

0.022	2019-05-01
	* Improved POD documentation
	* Fixed name in n50_algorithm.pl

0.021	2019-05-01
	* Improved POD documentation
	* Added a demo implementation of n50.pl using Proch::N50, keeping n50_algorithm.pl that does the same but without Proch::N50

0.02	2019-04-30
	* Added this Changelog
	* Removed unnecessary variables
	* Fixed POD documentation
	- TODO: Bareword filehandle: currently used as readfq() needs it

0.01	2019-04-30
	* Initial release on CPAN


n50/lib/Proch/N50.pm
package Proch::N50;
#ABSTRACT: a small module to calculate N50 (total size, and total number of sequences) for a FASTA or FASTQ file. It's easy to install, with minimal dependencies.

use 5.012;
use warnings;
my  $opt_digits = 2;
$Proch::N50::VERSION = '1.4.0';
use File::Spec;
use JSON::PP;
use FASTX::Reader;
use File::Basename;
use Exporter qw(import);

our @EXPORT = qw(getStats getN50 jsonStats);

=head1 SYNOPSIS

  use Proch::N50 qw(getStats getN50);
  my $filepath = '/path/to/assembly.fasta';

  # Get N50 only: getN50(file) will return an integer
  print "N50 only:\t", getN50($filepath), "\n";

  # Full stats
  my $seq_stats = getStats($filepath);
  print Data::Dumper->Dump( [ $seq_stats ], [ qw(*FASTA_stats) ] );
  # Will print:
  # %FASTA_stats = (
  #               'N50' => 65,
  #               'N75' => 50,
  #               'N90' => 4,
  #               'min' => 4,
  #               'max' => 65,
  #               'dirname' => 'data',
  #               'auN' => 45.02112,
  #               'size' => 130,
  #               'seqs' => 6,
  #               'filename' => 'test.fa',
  #               'status' => 1
  #             );

  # Get also a JSON object
  my $seq_stats_with_JSON = getStats($filepath, 'JSON');
  print $seq_stats_with_JSON->{json}, "\n";
  # Will print:
  # {
  #    "status" : 1,
  #    "seqs" : 6,
  #    <...>
  #    "filename" : "small_test.fa",
  #    "N50" : 65,
  # }
  # Directly ask for the JSON object only:
  my $json = jsonStats($filepath);
  print $json;

=head1 METHODS

=head2 getN50(filepath)

This function returns the N50 for a FASTA/FASTQ file given, or 0 in case of error(s).

=head2 getStats(filepath, alsoJSON)

Calculates N50 and basic stats for <filepath>. Returns also JSON if invoked
with a second parameter.
This function return a hash reporting:

=over 4

=item I<size> (int)

total number of bp in the files

=item I<N50>, I<N75>, I<N90> (int)

the actual N50, N75, and N90 metrices

=item I<auN> (float)

the area under the Nx curve, as described in L<https://lh3.github.io/2020/04/08/a-new-metric-on-assembly-contiguity>.
Returs with 5 decimal digits.

=item I<min> (int)

Minimum length observed in FASTA/Q file


=item I<max> (int)

Maximum length observed in FASTA/Q file


=item I<seqs> (int)

total number of sequences in the files


=item I<filename> (string)

file basename of the input file


=item I<dirname> (string)

name of the directory containing the input file (as received)

=item I<path> (string)

name of the directory containing the input file (resolved to its absolute path)

=back

=over 4

=item I<json> (string: JSON pretty printed)

(pretty printed) JSON string of the object (only if JSON is installed)

=back

=head2 jsonStats(filepath)

Returns the JSON string with basic stats (same as $result->{json} from I<getStats>(File, JSON)).
Requires JSON::PP installed.

=head2 _n50fromHash(hash, totalsize)

This is an internal helper subroutine that perform the actual N50 calculation, hence its addition
to the documentation.
Expects the reference to an hash of sizes C<$size{SIZE} = COUNT> and the total sum of sizes obtained
parsing the sequences file.
Returns N50, min and max lengths.

=head1 Dependencies

=head2 Module (N50.pm)

=over 4

=item L<FASTX::Reader> (required)

=item L<JSON::PP>, <File::Basename> (core modules)

=back

=head2 Implementation (n50.pl)

=over 4

=item L<Term::ANSIColor>

=back

=over 4

=item L<JSON>

(optional) when using C<--format JSON>

=back

=over 4

=item L<Text::ASCIITable>

(optional) when using C<--format screen>. This might be substituted by a different module in the future.

=back


=head1 SUPPORT

SeqFu is a compiled suite of utilities that includes a B<seqfu stats> module. 
SeqFu is currently the ideal choice that can replace the C<n50> program.

If you are interested in contributing to the development of this module, or
in reporting bugs, please refer to the legacy repository
L<https://github.com/quadram-institute-bioscience/seqfu/issues>.

=cut
sub getStats {
    # Parses a FASTA/FASTQ file and returns stats
    # Parameters:
    # * filename (Str)
    # * Also return JSON string (Bool)

    my ( $file, $wantJSON, $customN ) = @_;
    if (defined $customN and ($customN > 100 or $customN < 0) ) {
      die "[Proch::N50] Custom value must be 0 < x < 100\n";
    }
    my $answer;
    $answer->{status} = 1;
    $answer->{N50}    = undef;

    # Check file existence
# uncoverable condition right
    if ( !-e "$file" and $file ne '-' ) {
        $answer->{status}  = 0;
        $answer->{message} = "Unable to find <$file>";
    }



    # Return failed status if file not found or not readable
    if ( $answer->{status} == 0 ) {
        return $answer;
    }

    ##my @aux = undef;
    my $Reader;
    if ($file ne '-') {
       $Reader = FASTX::Reader->new({ filename => "$file" });
    } else {
       $Reader = FASTX::Reader->new({ filename => '{{STDIN}}' });
    }
    my %sizes;
    my ( $n, $slen ) = ( 0, 0 );

    # Parse FASTA/FASTQ file
    while ( my $seq = $Reader->getRead() ) {
        ++$n;
        my $size = length($seq->{seq});
        $slen += $size;
        $sizes{$size}++;
    }

    my ($n50, $min, $max, $auN, $n75, $n90, $nx);
    unless ($n) {
      ($n50, $min, $max, $auN, $n75, $n90, $nx) =
      (   0,    0,    0,    0,    0,    0,   0);
       say STDERR "[n50] WARNING: Not a sequence file: $file";
    } else {
      # Invokes core _n50fromHash() routine
      ($n50, $min, $max, $auN, $n75, $n90, $nx) = _n50fromHash( \%sizes, $slen, $customN );
    }
    my $basename = basename($file);

    $answer->{N50}      = $n50 + 0;
    $answer->{N75}      = $n75 + 0;
    $answer->{N90}      = $n90 + 0;
    if (defined $customN) {
      $answer->{Ne}       = $nx  + 0;
      $answer->{"N$customN"} = $nx  + 0;
    }

    $answer->{auN}      = sprintf("%.${opt_digits}f", $auN + 0);
    $answer->{min}      = $min + 0;
    $answer->{max}      = $max + 0;
    $answer->{seqs}     = $n;
    $answer->{size}     = $slen;
    $answer->{filename} = $basename;
    $answer->{dirname}  = dirname($file);
    $answer->{path   }  = File::Spec->rel2abs(dirname($file));

    # If JSON is required return JSON
    if ( defined $wantJSON and $wantJSON ) {

        my $json = JSON::PP->new->ascii->pretty->allow_nonref;
        my $pretty_printed = $json->encode( $answer );
        $answer->{json} = $pretty_printed;

    }
    return $answer;
}

sub _n50fromHash {
    # _n50fromHash(): calculate stats from hash of lengths
    #
    # Parameters:
    # * A hash of  key={contig_length} and value={no_contigs}
    # * Sum of all contigs sizes
    my ( $hash_ref, $total_size, $custom_n ) = @_;
    my $progressive_sum = 0;
    my $auN = 0;
    my $n50 = undef;
    my $n75 = undef;
    my $n90 = undef;
    my $nx  = undef;
    my @sorted_keys = sort { $a <=> $b } keys %{$hash_ref};

    # Added in v. 0.039
    my $max =  $sorted_keys[-1];
    my $min =  $sorted_keys[0] ;
    # N50 definition: https://en.wikipedia.org/wiki/N50_statistic
    # Was '>=' in my original implementation of N50. Now complies with 'seqkit'
    # N50 Calculation

    foreach my $s ( @sorted_keys ) {
        my $ctgs_length = $s * ${$hash_ref}{$s};
        $progressive_sum +=  $ctgs_length;


        $auN += ( $ctgs_length ) * ( $ctgs_length / $total_size);

       if ( !$n50 and $progressive_sum > ( $total_size * ((100 - 50) / 100) ) ) {
         $n50 = $s;
       }

       if ( !$n75 and $progressive_sum > ( $total_size * ((100 - 75) / 100) ) ) {
         $n75 = $s;
       }
       if ( !$n90 and $progressive_sum > ( $total_size * ((100 - 90) / 100) ) ) {
         $n90 = $s;
       }
       if ( !$nx and defined $custom_n) {
         $nx = $s if ( $progressive_sum > ( $total_size * ((100 - $custom_n) / 100) ));
       }
    }
    return ($n50, $min, $max, $auN, $n75, $n90, $nx);

}

sub getN50 {

    # Invokes the full getStats returning N50 or 0 in case of error;
    my ($file) = @_;
    my $stats = getStats($file);

# Verify status and return
# uncoverable branch false
    if ( $stats->{status} ) {
        return $stats->{N50};
    } else {
        return 0;
    }
}

sub jsonStats {
  my ($file) = @_;
  my $stats = getStats($file,  'JSON');

# Return JSON object if getStats() was able to reduce one
# uncoverable branch false
  if (defined $stats->{json}) {
    return $stats->{json}
  } else {
    # Return undef otherwise
    return undef;
  }
}


1;
